Squashing all previous pull requests for review

davidgarry1 opened the pull request:
As discussed: https://docs.google.com/document/d/1asE4yHGc2ltEYONIAdBsA3BOe0L_agr4p9NzqcFZFCA/

davidgarry1 committed:
Merge pull request #1 from google/master

Updating our fork with the master fixes to date errors and README typo

shuai9532 committed:
the core db function is available, but need to fix connection

shuai9532 committed:
updated connection code to db

shuai9532 committed:
Merge pull request #2 from jonathanloganmoran/shuaih

Shuaih connection code udpate

shuai9532 committed:
just little typo

shuai9532 committed:
updated connector, please read the note on top in the file

davidgarry1 committed:
Merge branch 'master' into shuaih

davidgarry1 committed:
nothing substantial

davidgarry1 committed:
Update .gitignore

davidgarry1 committed:
Update .gitignore

davidgarry1 committed:
username alphanumeric character validation

shuai9532 committed:
updated some comments

updated some comments for the constructor and explanation for the main method

shuai9532 committed:
Merge pull request #3 from jonathanloganmoran/shuaih

updated connector that talks to the database. The methods inside will be used to verify, update and create the user account

davidgarry1 committed:
Merge remote-tracking branch 'origin/master'

davidgarry1 committed:
Moved the character validation to a seperate class to be further expanded for other types of text validation.

davidgarry1 committed:
Moved the character validation to a seperate class to be further expanded for other types of text validation.

davidgarry1 committed:
Merge pull request #6 from jonathanloganmoran/davidBranch

David Branch

shuai9532 committed:
working databse

davidgarry1 committed:
Merge pull request #7 from jonathanloganmoran/shuaih

working database

davidgarry1 committed:
Delete Connector.java

davidgarry1 committed:
Added first tests for database, made security changes to Connector.java

davidgarry1 committed:
Update UuidsTest.java

davidgarry1 committed:
Merge pull request #8 from jonathanloganmoran/davidBranch

Added first tests for database, made security changes to Connector.java

shuai9532 committed:
jsbc connector added

shuai9532 committed:
Merge pull request #9 from jonathanloganmoran/shuaih

jdbc connector added

davidgarry1 committed:
Added first tests for database,

davidgarry1 committed:
Merge pull request #10 from jonathanloganmoran/davidBranch

No substantial changes -- syncrhonization

shuai9532 committed:
edited Connector.java and test.sh to incorperate jar into path

shuai9532 committed:
Merge pull request #11 from jonathanloganmoran/shuaih

edited Connector.java and edited test.sh to incorporate jar into path

shuai9532 committed:
added dropUsers() and printUsers()

shuai9532 committed:
added dropUsers() and printUsers()

shuai9532 committed:
added dropUsers() and printUsers()

shuai9532 committed:
added dropUsers() and printUsers()

shuai9532 committed:
edited some minor fixes

change to stderr for error messages

shuai9532 committed:
preparedStmt is used, Users are all returned in array,closeConnection is added

shuai9532 committed:
preparedStmt is used, Users are all returned in array,closeConnection is added

davidgarry1 committed:
Merge pull request #12 from jonathanloganmoran/shuaih

added dropUsers() and printUsers()

shuai9532 committed:
fixed try catch blocks and closure of objects

shuai9532 committed:
updated queries and minor changes

shuai9532 committed:
add a connection pool for faster and simoutaneous processing

shuai9532 committed:
jar file needed for connection pool

davidgarry1 committed:
Merge pull request #13 from jonathanloganmoran/shuaih

fixed try catch blocks and closure of objects

davidgarry1 committed:
Merging in fixes for database errors in master

shuai9532 committed:
change database to Databse

shuai9532 committed:
Update DatabaseTest.java

shuai9532 committed:
Merge pull request #14 from jonathanloganmoran/davidBranch

Merging in fixes for database errors in master

davidgarry1 committed:
naming

davidgarry1 committed:
Merge pull request #15 from jonathanloganmoran/davidBranch

naming of Database -> database

davidgarry1 committed:
Delete Connector.java

davidgarry1 committed:
Changing the test.sh script to include new jar file that is dependent.

davidgarry1 committed:
Merge pull request #16 from jonathanloganmoran/davidBranch

Changing the test.sh script to include new jar file that is dependent.

shuai9532 committed:
jar files needed for pool and log

shuai9532 committed:
manually updated

shuai9532 committed:
manually updated

shuai9532 committed:
manually adding files

shuai9532 committed:
resolving conflict

shuai9532 committed:
resolving conflict

shuai9532 committed:
conflicts

shuai9532 committed:
Merge branch 'master' of https://github.com/jonathanloganmoran/codeu_project_2017 into shuaih

shuai9532 committed:
resolved conflicts and fixed warning for file path

davidgarry1 committed:
Merge pull request #17 from jonathanloganmoran/shuaih

fixed conflicts and warnings for file path.

davidgarry1 committed:
holding changes to password verification

shuai9532 committed:
fix script for running simpleGui

shuai9532 committed:
Merge pull request #18 from jonathanloganmoran/shuaih

fix script for running simpleGui

davidgarry1 committed:
password updates

davidgarry1 committed:
Merge branches 'davidBranch' and 'master' of https://github.com/jonathanloganmoran/codeu_project_2017 into davidBranch

davidgarry1 committed:
Changes to show progress while testing is being conducted, as well as improved testing for database. Known error of logs being created outside of the bin folder from master branch still present.

davidgarry1 committed:
Client-Side verification through database of accounts.

davidgarry1 committed:
Delete chat_simple_gui_client_log.log

davidgarry1 committed:
Create .gitignore

davidgarry1 committed:
Delete codeu_project_2017.iml

davidgarry1 committed:
Delete misc.xml

davidgarry1 committed:
Delete modules.xml

davidgarry1 committed:
Delete vcs.xml

davidgarry1 committed:
Delete Connector.class

davidgarry1 committed:
Delete chat_server_log.log

davidgarry1 committed:
delete bin

davidgarry1 committed:
int

davidgarry1 committed:
Delete workspace.xml

shuai9532 committed:
Merge pull request #19 from jonathanloganmoran/davidBranch

Addition of client-side verification with remote database for permanent storage of accounts

davidgarry1 commented:
Has conflicts because Google updated their master with fixes for bugs discovered after we forked and began working on the repository.

dentearl commented on make.sh:
```diff
@@ -16,4 +16,4 @@
 mkdir -p bin
 javac -Xlint $(find * | grep "\\.java$") -d ./bin -sourcepath ./src -cp ./third_party/junit4.jar:./third_party/commons-dbcp2-2.1.1.jar:./third_party/mysql-connector-java-5.1.41-bin.jar:./third_party/commons-pool2-2.4.2.jar:./third_party/commons-logging-1.2.jar:./bin
-javac -Xlint $(find * | grep "\\.java$") -d ./bin -sourcepath ./test -cp ./third_party/junit4.jar:./third_party/commons-dbcp2-2.1.1.jar:./third_party/mysql-connector-java-5.1.41-bin.jar:./third_party/commons-pool2-2.4.2.jar:./third_party/commons-logging-1.2.jar:./bin
+javac -Xlint $(find * | grep "\\.java$") -d ./bin -sourcepath ./test -cp ./third_party/junit4.jar:./third_party/commons-dbcp2-2.1.1.jar:./third_party/mysql-connector-java-5.1.41-bin.jar:./third_party/commons-pool2-2.4.2.jar:./third_party/commons-logging-1.2.jar:./bin
```
End files with a newline please.

dentearl commented on src/codeu/chat/client/ClientUser.java:
```diff
@@ -132,7 +132,7 @@ public void updateUsers() {
   public static String getUserInfoString(User user) {
     return (user == null) ? "Null user" :
-        String.format(" User: %s\n   Id: %s\n   created: %s\n", user.name, user.id, user.creation);
+        String.format(" User: %s\n   Id: %s\n", user.name, user.id); // removed "    created: %s"
```
Unnecessary to leave in commented out code.

dentearl commented on src/codeu/chat/client/simplegui/UserPanel.java:
```diff
@@ -150,7 +157,15 @@ private void initialize() {
     this.add(currentPanel, currentPanelC);
-    //TextValidator validator = new TextValidator();
+
+    //load existing users from database
```
Format for comments is with a space between // and you text, proper capitalization and punctuation. So 
`// Load existing users from database.`

dentearl commented on src/codeu/chat/client/simplegui/UserPanel.java:
```diff
@@ -150,7 +157,15 @@ private void initialize() {
+    con = new Connector();
+    List<String> usersToAdd = con.getAllUsers();
+    for(String s : usersToAdd) {
+      clientContext.user.addUser(s);
```
nit (this was original code): Having clientContext.user a public field is gross (EJ item 14).

dentearl commented on src/codeu/chat/client/simplegui/UserPanel.java:
```diff
@@ -159,31 +174,80 @@ public void actionPerformed(ActionEvent e) {
-          userSignedInLabel.setText("Hello " + data);
+          JTextField username = new JTextField(data);
+          JTextField password = new JPasswordField();
+          Object[] message = {
```
Having this be a raw Object seems fishy, are you sure you don't want String? 

Also, you're showing the user's password in plain text?

dentearl commented on src/codeu/chat/client/simplegui/UserPanel.java:
```diff
@@ -159,31 +174,80 @@ public void actionPerformed(ActionEvent e) {
-              JOptionPane.ERROR_MESSAGE);
+        JTextField username = new JTextField();
+        JTextField password = new JPasswordField();
+        Object[] message = {
```
`String[] message = { ...`

dentearl commented on src/codeu/chat/client/simplegui/UserPanel.java:
```diff
@@ -159,31 +174,80 @@ public void actionPerformed(ActionEvent e) {
-              "User not created. Alphanumeric characters only, with no spaces. Please try again.",
-              "User Not Created",
-              JOptionPane.ERROR_MESSAGE);
+        JTextField username = new JTextField();
```
This whole section looks remarkably similar to the one above.

dentearl commented on src/codeu/chat/util/TextValidator.java:
```diff
@@ -9,6 +9,7 @@
 public final class TextValidator {
 
   private static final Pattern validUserNamePattern = Pattern.compile("[a-zA-Z0-9]+");
+  private static final Pattern validPasswordPattern = Pattern.compile("[a-zA-Z0-9]+");
```
`private static final` is effectively a constant and constants have names like such
`VALID_PASSWORD_PATTERN`

dentearl commented on src/database/Connector.java:
```diff
@@ -87,13 +87,13 @@ public boolean addAccount(String username, String password) {
       try (PreparedStatement dropAll = conn.prepareStatement(sqlDrop)) {
         dropAll.executeUpdate();
-        System.out.println("the table has been cleared");
+        //System.out.println("the table has been cleared");
```
As a rule, don't check in commented out code.

dentearl commented on src/database/Connector.java:
```diff
@@ -65,7 +65,7 @@ public Connector() {
-  public boolean addAccount(String username, String password) {
+  public synchronized boolean addAccount(String username, String password) {
 
     String sqlInsertAccount = "insert into "+tableName + "(username, password)" + "values(?,?)";
```
Space between binary operators and after commas, it's good practice to capitalize SQL commands
`String sqlInsertAccount = "INSERT INTO " + tableName + "(username, password)" + "values(?, ?)";`

dentearl commented on src/database/Connector.java:
```diff
@@ -124,16 +124,44 @@ public boolean verifyAccount(String username, String password) {
+   * verify if the account username input by users matches an account in the
+   * database
+   *
+   * @param username
```
The description is missing from param username.
`@param username name of the user account to verify.`

dentearl commented on src/database/Connector.java:
```diff
@@ -124,16 +124,44 @@ public boolean verifyAccount(String username, String password) {
+   */
+  public synchronized boolean verifyAccountExists(String username) {
+
+    String sqlSelectPassword = "select password from " + tableName + " where username = ?";
```
It seems sketchy to use a `select password` as a test of the user's existence. Is there something else you can use, something that isn't a security risk?

Formatting strings like this can often be stored as constants at the top of the class.

shuai9532 replied:
I think the method has been updated since last commit. This is in my code:
`public boolean addAccount(String username, String password);`
use usrname to get the password in the database, and then compare to the password passed in as a parameter to validate the account. In the future, the password will be encrypted, and I am currently looking into digesting the password @dentearl

dentearl replied:
(^-^)b

dentearl commented on src/database/Connector.java:
```diff
@@ -143,7 +171,7 @@ public boolean verifyAccount(String username, String password) {
-  public boolean deleteAccount(String username) {
+  public synchronized boolean deleteAccount(String username) {
 
     String sqlDeleteAccount = "DELETE  FROM " + tableName + " WHERE username = ?";
```
Double space between DELETE and FROM.

dentearl commented on test/codeu/chat/TestRunner.java:
```diff
@@ -17,23 +17,85 @@
-         System.out.println(failure.toString());
+
+  private static boolean loading = true;
+  private static int frameSpeed = 60;
```
Constants, `FRAME_SPEED`

dentearl commented on test/codeu/chat/TestRunner.java:
```diff
@@ -17,23 +17,85 @@
+          }
+          String results = "\rResult:                     "
+              + (result.wasSuccessful() ? "\n\tPassed\n"
+              : "\n\tFailed " + numFailures + " test(s):\n");
```
`String.format()` might be useful here, along with a simple conditional statement for
`if (result.wasSeccussful()) {...`

dentearl reviewed (changes requested):
OK set of changes.

dentearl reviewed (commented):
Pull 3ef7b18,
The commit message for this pull request is not descriptive at all.

dentearl commented on src/database/Connector.java:
```diff
@@ -0,0 +1,192 @@
+  private static final String dbName = "CodeU_2017DB";
+  private static final String hostname = "ec2-176-34-225-252.eu-west-1.compute.amazonaws.com";
+  private static final String DBusername = "group34";
+  private static final String DBpassword = "codeu2017";
```
Storing a username and password in a source-controlled file is a security risk. One way to get around problems like this is to have a file you store on disk that your .gitignore knows to ignore and you read the credentials from that file (both username and pass). Each developer will have to know those credentials / create those files but otherwise you're exposing yourself.

dentearl commented on src/database/Connector.java:
```diff
@@ -0,0 +1,192 @@
+  public List<String> getAllUsers() {
+
+    String sqlSelectUsername = "select username from " + tableName;
+    LinkedList<String> userNames = new LinkedList<>();
```
`List<String> userNames = newLinkedList<>();`
Effective Java item 52, refer to objects by their interfaces.

dentearl commented on src/database/Connector.java:
```diff
@@ -0,0 +1,192 @@
+            // get the stored password from database
+            String passwordInDB = resultPassword.getString(1);
+            // password does not match
+            if (!passwordInDB.equals(password)) {
```
Are you guys storing the password in plaintext? If so this is a feature to improve :)

dentearl commented on test/codeu/chat/server/DatabaseTest.java:
```diff
@@ -5,15 +5,17 @@
   @Test
   public void testAddAndDeleteUser() {
-    System.out.println("DATABASE TEST RUN");
+    Connector con = new Connector();
```
There's an `@Before` annotation that I believe can be invoked on the `setup()` method — this will run before each test. There's also a `@After` for cleanup / teardown operations.

dentearl commented on src/codeu/chat/client/ClientUser.java:
```diff
@@ -49,10 +49,11 @@ static public boolean isValidName(String userName) {
     if (userName.length() == 0) {
       clean = false;
+    //if alphanumeric
+    } else if(!userName.matches("[a-zA-Z0-9]+")) {
```
Don't you guys already have a compiled Pattern somewhere with this username pattern? Consolidating the pattern in one location will allow you to seamlessly update if you decide to allow more interesting characters,

dentearl commented on src/codeu/chat/client/ClientUser.java:
```diff
@@ -49,10 +49,11 @@ static public boolean isValidName(String userName) {
     boolean clean = true;
     if (userName.length() == 0) {
       clean = false;
+    //if alphanumeric
```
Omit comments that don't help explain the code. And when you do comment, use proper capitalization and punctuation, so
`// If alphanumeric.`

dentearl commented on src/codeu/chat/client/simplegui/UserPanel.java:
```diff
@@ -164,8 +164,16 @@ public void actionPerformed(ActionEvent e) {
-          clientContext.user.addUser(s);
-          UserPanel.this.getAllUsers(listModel);
+          //if alphanumeric
+          if(s.matches("[a-zA-Z0-9]+")) {
```
Patterns stored across classes that need to be kept in sync can go out of sync easily. Maybe a central common constant class?

dentearl commented on src/DBConnector/Connector.java:
```diff
@@ -19,20 +20,46 @@
    private static final String DBpassword = "codeu2017";
 
+   //the constructor can only create once per run.
+   private
```
The keyword `private` and the type `PreparedStatement` should be on the same line together. Break the line into two for the first variable that crosses the 120 column mark.

dentearl commented on src/DBConnector/Connector.java:
```diff
@@ -44,25 +71,36 @@ public Connector() {
-
-    String sql = "select username from " + tableName;
-
+  public String[] getAllUsers() {
```
Return a `List<String>`.
Effective Java item 25, prefer lists to arrays.

dentearl commented on src/DBConnector/Connector.java:
```diff
@@ -44,25 +71,36 @@ public Connector() {
+      resultSize = selectCount.executeQuery();
+      resultSize.next();
+      int size = resultSize.getInt(1);
+      String[] userNames = new String[size];
```
`List<String> userNames = new ArrayList<>();`
I wouldn't worry about setting capacity on your list at this point.

dentearl commented on src/DBConnector/Connector.java:
```diff
@@ -75,49 +113,56 @@ public void printAllUsers() {
 
-    /* string constructed for database */
-    String sqlAdd = "insert into "+tableName + "(username, password)" + "values(" + "'" + username
-        + "'" + "," + "'" + password + "'" + ")";
```
This would read more easily if you used a `Joiner.on("'").join(fields)`
https://google.github.io/guava/releases/16.0/api/docs/com/google/common/base/Joiner.html

dentearl replied:
And maybe even set up a `private static final String QUOTE = "'";`

dentearl commented on src/DBConnector/Connector.java:
```diff
@@ -75,49 +113,56 @@ public void printAllUsers() {
       // if the table is empty, we directly insert the new account
       if (size == 0) {
-        myStmt.executeUpdate(sqlAdd);
+        add.setString(1,username);
```
Space after comma.

dentearl commented on src/DBConnector/Connector.java:
```diff
@@ -19,20 +20,46 @@
+      selectUsername = myConn.prepareStatement("select username from "+tableName+" where username = ?");
+
+      /*for addUser():*/
+      add = myConn.prepareStatement("insert into "+tableName + "(username, password)" + "values(?,?)");
```
This variable name is a bit misleading. Maybe `addStatement`? Just seeing `add` referenced later I found it confusing, especially since `add` is a common method for the `Collection` interface.

dentearl commented on src/DBConnector/Connector.java:
```diff
@@ -75,49 +113,56 @@ public void printAllUsers() {
       // the newly created username does not exist in the database, then add it to the database
-      myStmt.executeUpdate(sqlAdd);
+      add.setString(1,username);
+      add.setString(2,password);
```
If you're feeling industrious, this chain of calls makes the case for a Builder for this statement — using `setString(` and an index value is error prone.

shuai9532 replied:
I am not very sure what it means here @dentearl

dentearl replied:
Ah, I'm saying it'd be slicker to be like 
```Java
CodeUSqlStatementBuilder.builder()
  .addUser()
  .withUsername(username)
  .withPassword(password)
  .build()
  .execute();
```

Or something similar. This allows you to have validation checking in your build() method and then the actual Connector call in execute(). It's just cleaner syntax and better input validation. It's more work though, which is why I'm saying it's only if you're feeling industrious ;)

dentearl commented on src/DBConnector/Connector.java:
```diff
@@ -227,15 +283,36 @@ public boolean deleteAccount(String username) {
+    } catch (SQLException e) {
+      e.printStackTrace();
+    }
+    try{myConn.close();}catch (SQLException e) {e.printStackTrace();}
```
Expand this out to be on multiple lines.

dentearl commented on src/codeu/chat/client/ClientUser.java:
```diff
@@ -31,7 +31,7 @@
   private final Controller controller;
   private final View view;
-  private final TextValidator validator = new TextValidator();
+  //private final TextValidator validator = new TextValidator();
```
Don't check in commented-out code.

dentearl commented on src/codeu/chat/client/ClientUser.java:
```diff
@@ -143,4 +143,4 @@ public String showUserInfo(String uname) {
     System.out.println(getUserInfoString(user));
   }
-}
+}
```
Always end files with newlines.

dentearl commented on test/codeu/chat/server/DatabaseTest.java:
```diff
@@ -0,0 +1,29 @@
+
+public final class DatabaseTest {
+
+  private static final DBConnector.Connector con = new DBConnector.Connector();
```
`@Before` with `setUp()` and `@After` with `tearDown()` would probably be useful with your connector.

dentearl commented on src/codeu/chat/util/TextValidator.java:
```diff
@@ -0,0 +1,22 @@
+ */
+public final class TextValidator {
+
+  private static final Pattern validUserNamePattern = Pattern.compile("[a-zA-Z0-9]+");
```
`VALID_USER_NAME_PATTERN` per style guide.

dentearl commented on src/codeu/chat/util/TextValidator.java:
```diff
@@ -0,0 +1,22 @@
+  private static final Pattern validUserNamePattern = Pattern.compile("[a-zA-Z0-9]+");
+
+  public static boolean isValidUserName(String username) {
+    if (username == null) {
```
You could consolidate this conditional a bit with 
`com.google.common.base.Strings Strings.isNullOrEmpty()`

dentearl commented on src/codeu/chat/util/TextValidator.java:
```diff
@@ -0,0 +1,22 @@
+      return false;
+    }
+  }
+}
```
End with newline

shuai9532 committed:
fix according to Dent

shuai9532 committed:
added Guava lib, changed style, added encryption, handled screte

shuai9532 committed:
fixed style after the review, added encryption

davidgarry1 committed:
Merge branch 'master' into shuaih

shuai9532 committed:
delete error message, added ddelete test users

shuai9532 committed:
updated changes from review

shuai9532 committed:
minor style change

davidgarry1 committed:
Merge pull request #20 from jonathanloganmoran/shuaih

fixed style after the review, added encryption

davidgarry1 committed:
Formatting changes, added 1 method to Database for verifying an account exists.

Known issue with @shuai code: the DB fails 1 test, updatePassword()

davidgarry1 committed:
Formatting changes, added 1 method to Database for verifying an account exists.

Known issue with @shuai code: the DB fails 1 test, updatePassword()

davidgarry1 committed:
Removed the dropDatabase test, we know it works but it deletes all users,
does not allow testing that is non-disruptive.

Known issue with @shuai code: the DB fails 1 test, updatePassword()

davidgarry1 committed:
Fixed issue in VerifyAccount()

Known issue with @shuai code: the DB fails 1 test, updatePassword()

davidgarry1 committed:
Merge pull request #21 from jonathanloganmoran/davidBranch

Formatting changes, added DB method to verify account exists (no PW)

davidgarry1 committed:
Fixed issue in VerifyAccount()

Known issue with @shuai code: the DB fails 1 test, updatePassword()

davidgarry1 committed:
Merge pull request #22 from jonathanloganmoran/davidBranch

Fixed issue in VerifyAccount()

shuai9532 committed:
added exceptions, added new line, fixed digits selection and some style problems

shuai9532 committed:
databseTest commited temporarily

shuai9532 committed:
temporary commit

shuai9532 committed:
dasd

shuai9532 committed:
fixing verify

shuai9532 committed:
fixed database account verify after update and some minor changes

shuai9532 committed:
style fix

davidgarry1 committed:
Merge pull request #23 from jonathanloganmoran/shuaih

this request fixed the issue of verifying account after updating the password

shuai9532 committed:
deleted StoreAcess

shuai9532 committed:
added tests,conversation&message added, style changed

shuai9532 committed:
Update Connector.java

shuai9532 committed:
Update TestConnector.java

shuai9532 committed:
Update TestConnector.java

shuai9532 committed:
Update history.sql

shuai9532 committed:
Update DatabaseTest.java

davidgarry1 committed:
Merge pull request #24 from jonathanloganmoran/shuaih

updated some function regarding conversation and message

shuai9532 committed:
fixed some warnings and add serial to the created exception

shuai9532 committed:
Merge branch 'shuaih' of https://github.com/jonathanloganmoran/codeu_project_2017 into shuaih

shuai9532 committed:
minor style update

shuai9532 committed:
Merge branch 'master' of https://github.com/jonathanloganmoran/codeu_project_2017 into shuaih

shuai9532 committed:
Merge pull request #25 from jonathanloganmoran/shuaih

warning and error of used import fixed

davidgarry1 committed:
Addition of web portal for CodeU chat app
